### PR TITLE
Calculate response boundaries explicitly

### PR DESCRIPTION
**Pull request by arichiardi**
Change the way the response is calculated, not by `search-forward` the buffer used for redirection but by parsing the raw response. Taking into consideration the optional regexps, a list is calculated with beginning, end and prompt position (the logic is now in inf-clojure--string-boundaries, which can be unit tested).

This implicitly solves a hidden bug in inf-clojure--process-response, which was not able to identify the prompt in some cases because of the many search-forward.

Finally, this new information is logged so that bugs can be detected more easily.

- [X] The commits are consistent with our [contribution guidelines][1]
- [X] The new code is not generating bytecode or `M-x checkdoc` warnings

### CODE REVIEW

File: inf-clojure.el [hunk 1]
```diff
@@ -1002,17 +1013,16 @@ the `inf-clojure-prompt`."
...
-                      (point-min)))
-             (end (or (when end-string
-                        (search-forward end-string nil t))
-                      (point-max)))
-             (prompt (when (search-forward inf-clojure-prompt nil t)
-                       (match-beginning 0)))
-             (buffer-string (buffer-substring-no-properties beg (or prompt end))))
-        (inf-clojure--log-string buffer-string "<-RES----")
-        buffer-string))))
+      (let* ((buffer-string (buffer-substring-no-properties (point-min) (point-max)))
```
Comment (bbatsov) [inline inf-clojure.el]: What is `buffer-string` supposed to be here?

File: inf-clojure.el [hunk 2]
```diff
@@ -1002,17 +1013,16 @@ the `inf-clojure-prompt`."
...
+             (boundaries (inf-clojure--string-boundaries buffer-string inf-clojure-prompt beg-regexp end-regexp))
+             (beg-pos (car boundaries))
+             (end-pos (car (cdr boundaries)))
+             (prompt-pos (car (cdr (cdr boundaries))))
+             (response-string (substring buffer-string beg-pos (min end-pos prompt-pos))))
+        (inf-clojure--log-string buffer-string "<-BUF----")
```
Comment (bbatsov) [inline inf-clojure.el]: I think those markers you should would be understandable just to you. :-)

Comment (bbatsov) [inline inf-clojure.el]: I also wonder who'd care about them most of the time.

**Comment by bbatsov** [comment]
Parsing which string?

File: inf-clojure.el (hunk 1 above)
Comment (arichiardi) [inline inf-clojure.el]: uhm you mean you don't like the name? Naming is hard..

File: inf-clojure.el (hunk 2 above)
Comment (arichiardi) [inline inf-clojure.el]: It is a debugging feature, we can of course improve the naming and the markings, proposals are welcome!

**Comment by arichiardi** [comment]
Improved description of the PR a bit.

File: inf-clojure.el (hunk 2 above)
Comment (bbatsov) [inline inf-clojure.el]: I really don't think this information would be useful to anyone 99% of the time, so I'm wary of logging it by default.

Comment (arichiardi) [inline inf-clojure.el]: Keep in mind that it is logged only when `inf-clojure-log-activity` is `t`. It is not even a `defcustom` so you really are a dev/contributor if you are using this info. And the more info you have the more you debug nicely. Having said that, what would you remove? I am fine with putting the few lines I need when I debug.

Comment (bbatsov) [inline inf-clojure.el]: Yeah, but that's not really process activity. It's just some random debug output that's completely useless unless you're debugging exactly this part of the code. I just think that this activity log should consistent and not a dumping ground for every debugging output one might decide to use. (and I usually debug with `edebug` so I'm not fond of dumping output like this in general)

Comment (arichiardi) [inline inf-clojure.el]: I have tried with `edebug` to debug this kind of process output. No success. Maybe I was doing something wrong. In any case I can take it away. That is the core of the plugin and you need it like bread.

Comment (bbatsov) [inline inf-clojure.el]: Well, OK. I still don't like it, but I'm also way to busy to discuss things of such little importance is great details.

**Pull request merged**